openmc/deplete/flux_operator.py

```diff
@@ -0,0 +1,594 @@
+        The depletion chain information necessary to form matrices and tallies.
+    reaction_rates : openmc.deplete.ReactionRates
+        Reaction rates from the last operator step.
+    burnable_mats : list of str
+        All burnable material IDs
+    heavy_metal : float
+        Initial heavy metal inventory [g]
+    local_mats : list of str
+        All burnable material IDs being managed by a single process
+    prev_res : Results or None
+        Results from a previous depletion calculation. ``None`` if no
+        results are to be used.
+
+       """
+
+    def __init__(self,
+                 materials,
+                 micro_xs,
+                 chain_file,
+                 keff=None,
+                 normalization_mode = 'constant-flux',
+                 fission_q=None,
+                 prev_results=None,
+                 reduce_chain=False,
+                 reduce_chain_level=None,
+                 fission_yield_opts=None):
+        # Validate micro-xs parameters
+        check_type('materials', materials, openmc.Materials)
+        check_type('micro_xs', micro_xs, pd.DataFrame)
+        if keff is not None:
+            check_type('keff', keff, tuple, float)
+            keff = ufloat(*keff)
+
+        self._keff = keff
+
+        helper_kwargs = dict()
+        helper_kwargs = {'normalization_mode': normalization_mode,
+                         'fission_yield_opts': fission_yield_opts}
+
+        super().__init__(
+            materials,
+            micro_xs,
+            chain_file,
+            prev_results,
+            fission_q=fission_q,
+            helper_kwargs=helper_kwargs,
+            reduce_chain=reduce_chain,
+            reduce_chain_level=reduce_chain_level)
+
+    @classmethod
+    def from_nuclides(cls, volume, nuclides, micro_xs,
+                      chain_file,
+                      keff=None,
+                      normalization_mode='constant-flux',
+                      fission_q=None,
+                      prev_results=None,
+                      reduce_chain=False,
+                      reduce_chain_level=None,
+                      fission_yield_opts=None):
+        """
+        Alternate constructor from a dictionary of nuclide concentrations
+
+        volume : float
+            Volume of the material being depleted in [cm^3]
+        nuclides : dict of str to float
+            ,Dictionary with nuclide names as keys and nuclide concentrations as
+            values. Nuclide concentration units are [atom/cm^3].
+        micro_xs : pandas.DataFrame
+            DataFrame with nuclides names as index and microscopic cross section
+            data in the columns. Cross section units are [cm^-2].
+        chain_file : str
+            Path to the depletion chain XML file.
+        keff : 2-tuple of float, optional
+           keff eigenvalue and uncertainty from transport calculation.
+           Default is None.
+        normalization_mode : {"constant-power", "constant-flux"}
```

Comment:
So after skimming that paper, I have a few points of understanding I'd like to reach consensus on:
- the `constant-power` normalization mode _is_ constant power but with the assumptions that the flux is the same at the beginning of the depletion step as at the end of the depletion step, which will give a flux that is different from the "true" value, which may not be a very good assumption.
- to get better results for the flux calculated from the constant power, we need to perform substeps to converge to a value of flux that is closer to the "true" value
- Because we can only accept a single `DataFrame` for the one-group cross-sections, even if we wanted to do this kind of substep renormalization, we couldn't because that would require us to calculate new cross sections for each substep.

We keep running into this issue of having to recalculate the cross sections. Implementing this kind of functionality (and additionally all the extra cool things you can do with it, like this substep renormalization) would be an interesting project!

I think the `constant-fission` name is fine, but we probably should try to be consistent with the rest of the `deplete` module and revert it to `source-rate`.

Reply:
> * with the assumptions that the flux is the same at the beginning of the depletion step as at the end of the depletion step

That _is_ the constant flux assumption :smile: The only difference really is how you got that constant flux in the first place — specifying it explicitly or deducing it from the power.

> * we need to perform substeps to converge to a value of flux

Yes, that is the only way I could see to adjust the flux so that the power is not decreasing over the full step.

> * Because we can only accept a single `DataFrame` for the one-group cross-sections, even if we wanted to do this kind of substep renormalization, we couldn't

Cross sections are only one piece of the puzzle. We _would_ have updated number densities, so even if the cross sections are constant, you need a higher flux to account for the decrease in the density of fissionable nuclides to maintain constant power. I think that's actually the primary effect -- I wouldn't expect the micros to change much over a single step since the spectrum is probably about the same.

> Implementing this kind of functionality ... would be an interesting project!

Agreed!